src/devices/bus/sg1000_exp/sp400.cpp

```diff
@@ -0,0 +1,210 @@
+// license: BSD-3-Clause
+// copyright-holders: Charles MacDonald, Devin Hill, Fabio Dalla Libera
+
+/*
+  license:BSD-3-Clause
+  Charles MacDonald
+  Devin Hill
+  Fabio Dalla Libera
+
+  6805 dumped using https://github.com/charlesmacd/HD6805_Reader
+
+  IC list:
+
+  * M5L8035LP (Mitsubishi clone of I8035)
+
+  P1 is the deserialized output data
+  P2.4 is the input serial data
+  P2.5 is busy (output to the computer)
+  P2.6 is the output strobe
+  P2.7 is busy (input from the 6805)
+
+
+  * HD6805V (Hitachi chip, pin compatible with MC6805P2, with 4K of ROM)
+
+  port B connected to the motors
+  port C connected to the pen and misc signals
+  port D connected to M5L8035LP parallel data
+
+*/
+
+
+#include "emu.h"
+#include "sp400.h"
+
+
+sp400_device::sp400_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock):
+	device_t(mconfig, SP400_PRINTER, tag, owner, clock),
+	device_sk1100_printer_port_interface(mconfig, *this),
+	m_dsercpu(*this, "dsercpu"),
+	m_motcpu(*this, "motcpu"),
+	m_data(0),
+	m_busy(1),
+	m_dserdata(0xFF),
+	m_dserstrobe(1),
+	m_motbusy(0),
+	m_motPenUp(1),
+	m_motPenDown(1),
+	m_plotter(*this, "plotter"),
+	m_frontbuttons(*this, "FRONTBUTTONS"),
+	m_misc(*this, "MISC")
+{
+}
+
+void sp400_device::device_start()
+{
+}
+
+void sp400_device::dser_map(address_map &map)
+{
+	map(0x0000, 0x0fff).rom();
+}
+
+void sp400_device::dser_p1_w(uint8_t v)
+{
+	m_dserdata = v;
+}
+
+void sp400_device::dser_p2_w(uint8_t v)
+{
+	m_busy = BIT(v, 5);
+	m_dserstrobe = BIT(v, 6);
+	m_motcpu->set_input_line(M6805_IRQ_LINE, m_dserstrobe);
+}
+
+uint8_t sp400_device::dser_p2_r()
+{
+	return
+		(m_motbusy<<7)|
+		(m_dserstrobe<<6)|
+		(m_busy<<5)|
+		(m_data<<4)|
+		(1<<3)|
+		(1<<2)|
+		(1<<1)|
+		(1<<0);
+}
+
+uint8_t sp400_device::mot_pa_r()
+{
+	return
+		(0<<0)|
+		(0<<1)|
+		(m_frontbuttons->read()<<2)| // bits 2,3 and 4
+		(0<<5)|
+		(0<<6);
+}
+
+void sp400_device::mot_pb_w(uint8_t v)
+{
+	m_plotter->update_motors(v&0x0F, (v>>4)&0x0F);
+}
+
+
+uint8_t sp400_device::mot_pc_r()
+{
+	return
+		(m_motPenUp<<0)|
+		(m_motPenDown<<1)|
+		(1<<2)|
+		(m_plotter->get_reedswitch_state()<<3)|
+		(m_motbusy<<4)|
+		(1<<5)|(1<<6)|(1<<7); //unconfirmed
+}
+
+void sp400_device::mot_pc_w(uint8_t v)
+{
+	m_motPenUp = BIT(v, 0);
+	m_motPenDown = BIT (v, 1);
+	m_motbusy = BIT(v, 4);
+	penZ();
+}
+
+uint8_t sp400_device::mot_pd_r()
+{
+	return sp400_device::m_dserdata;
+}
+
+void sp400_device::penZ()
```

Comment:
Not sure I like this function name, doesn't sound explicit off the bat unless it's some technical term from documentation?
Also it's used only once, is it supposed to be called again in a (missing) `device_reset` fn to put it in known state at boot?

Reply:
I changed it update_pen_state. I assume the plotter class already initializes the pen position.